### PR TITLE
Implement strstr

### PR DESCRIPTION
**Pull request by liutingjieni**
Add a new built-in func like strstr()

Add the strstr () function to make fuzzy comparisons when there is no way to confirm the absolute path of the file or the full name of the process or

usage:
`sudo bpftrace -e 't:syscalls:sys_enter_execve /strstr(comm,"sh") == 0/ { printf("%s called %s\n", comm, str(args->filename));}'`

Problem background: https://github.com/iovisor/bpftrace/issues/2306

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by liutingjieni** [comment]
Pr was here before ( [add strstr() by liutingjieni · Pull Request #2344 · iovisor/bpftrace](https://github.com/iovisor/bpftrace/pull/2344) ). Because of the conflict with master, so based on the latest master commit changes

**Comment by liutingjieni** [comment]
I see a check error, but I don't quite understand。what do I need to do now?

**Comment by viktormalik** [comment]
> Pr was here before ( [add strstr() by liutingjieni · Pull Request #2344 · iovisor/bpftrace](https://github.com/iovisor/bpftrace/pull/2344) ). Because of the conflict with master, so based on the latest master commit changes

In future, please keep updating the same PR (by force-push into the branch), opening a new one after each rebase is not a good idea.

> I see a check error, but I don't quite understand。what do I need to do now?

Your new test is failing with segmentation fault, so you'll have to debug. If you want to reproduce the CI environment, you can build and use the prepared Docker images. Check `.github/workflows/ci.yml` for exact Docker commands.

**Comment by liutingjieni** [comment]
I found that the error is build_test (LLVM 15 Release). I tried to remove the commit about it( Latest Commit) and I found that it works.

**Comment by viktormalik** [comment]
> I found that the error is build_test (LLVM 15 Release). I tried to remove the commit about it( Latest Commit) and I found that it works.

Which could mean that your code doesn't work with LLVM 15 (very likely it doesn't work with opaque pointers). Please check and fix that.

**Comment by liutingjieni** [comment]
all checks have passed. what else do I need to do now?

**Comment by viktormalik** [comment]
This is a terrible solution. You can't just arbitrarily revert a commit from master if it doesn't suit you.

You have to make your code work with [opaque pointers](https://llvm.org/docs/OpaquePointers.html). Once you do that, the checks will pass.

**Comment by liutingjieni** [comment]
> > This is a terrible solution. You can't just arbitrarily revert a commit from master if it doesn't suit you.
> > You have to make your code work with [opaque pointers](https://llvm.org/docs/OpaquePointers.html). Once you do that, the checks will pass.
> 
> I have changed my code with opaque pointers. The check for llvm15 has passed. The last two commit hashes from master have changed due to cherry-pick, but the code content has not changed.

**Comment by liutingjieni** [comment]
I want the code to be checked

**Comment by viktormalik** [comment]
> I have changed my code with opaque pointers. The check for llvm15 has passed. The last two commit hashes from master have changed due to cherry-pick, but the code content has not changed.

Ah, right, sorry. I'll review when I have time.

File: src/ast/irbuilderbpf.cpp
Comment (BurntBrunch) [inline src/ast/irbuilderbpf.cpp:870]: We should handle both strings being literal and not even emit the branches below.

File: src/ast/irbuilderbpf.cpp
Comment (BurntBrunch) [inline src/ast/irbuilderbpf.cpp:905]: Basic block names would ideally be unique. Can you append the idx in the first string here?

File: src/ast/irbuilderbpf.cpp
Comment (BurntBrunch) [inline src/ast/irbuilderbpf.cpp:908]: Same

**Review by BurntBrunch — COMMENTED** [review]
This looks fine for what it is but I have a meta question - why are we unrolling the loop?

For a char[32] vs char[32] call, this emits ~500 instructions.
```
> src/bpftrace -vv -e 't:mac80211:wake_queue {if (strstr(ar
gs->wiphy_name, args->wiphy_name)) {exit()}}'
INFO: node count: 13
Attaching 1 probe...

Program ID: 1422

The verifier log: 
processed 522 insns (limit 1000000) max_states_per_insn 0 total_states 32 peak_states 32 mark_read 24
```

Given that bpf has had backward branches for ages, why don't we just emit this as an actual loop with known bounds and let the verifier do the hard work? It would dramatically improve complexity while potentially also allowing probed strings to work (by emitting the loop against the probe size limit or even against the runtime string length, assuming we can satisfy the verifier).

File: tests/semantic_analyser.cpp
Comment (viktormalik) [inline tests/semantic_analyser.cpp:2046]: Isn't this basically a duplicate of line 2042?

File: man/adoc/bpftrace.adoc
Comment (fbs) [inline man/adoc/bpftrace.adoc:1639]: maybe call them needle/haystack?

File: man/adoc/bpftrace.adoc
Comment (fbs) [inline man/adoc/bpftrace.adoc:1642]: I feel like this was discussed before but I'm not convinced by the `strstr` name.  This function doesn't actually behave the same as `strstr` does in the C libs.

Maybe `strcontains` is a better name?

From the `strstr` manpages:

>      If needle is an empty string, haystack is returned; if needle occurs nowhere in haystack, NULL is returned; otherwise a pointer to the first character of the first occurrence of needle is returned.

This one behaves completely different, it returns 0 *when found*.

**Comment by fbs** [comment]
> This looks fine for what it is but I have a meta question - why are we unrolling the loop?
> 
> For a char[32] vs char[32] call, this emits ~500 instructions.
> 
> ```
> > src/bpftrace -vv -e 't:mac80211:wake_queue {if (strstr(ar
> gs->wiphy_name, args->wiphy_name)) {exit()}}'
> INFO: node count: 13
> Attaching 1 probe...
> 
> Program ID: 1422
> 
> The verifier log: 
> processed 522 insns (limit 1000000) max_states_per_insn 0 total_states 32 peak_states 32 mark_read 24
> ```
> 
> Given that bpf has had backward branches for ages, why don't we just emit this as an actual loop with known bounds and let the verifier do the hard work? It would dramatically improve complexity while potentially also allowing probed strings to work (by emitting the loop against the probe size limit or even against the runtime string length, assuming we can satisfy the verifier).

Agree but maybe we shouldn't block on it.

For a future improvement it would also be nice if we compare more bytes at once

File: docs/reference_guide.md
Comment (viktormalik) [inline docs/reference_guide.md:2893]: The description doesn't fit the parameter names.

File: man/adoc/bpftrace.adoc
Comment (viktormalik) [inline man/adoc/bpftrace.adoc:1641]: Same as above.

**Review by viktormalik — CHANGES_REQUESTED** [review]
Plus there are several unresolved comments (and some that were incorrectly resolved) from the previous reviews.

File: man/adoc/bpftrace.adoc
Comment (fbs) [reply man/adoc/bpftrace.adoc:1642]: @BurntBrunch @viktormalik  how do you feel about this rename? it got implemented now but I wanted to start a discussion :)

Comment (viktormalik) [reply man/adoc/bpftrace.adoc:1642]: I think that it's fine, although I'd expect it to return 1 if the substring is found, so that you can test for the positive case instead of the negative one when using it in a condition.

Another option would be to introduce a new operator for this (`in`?) but that's a bigger change and would have to be discussed more, I guess.

Comment (viktormalik) [reply man/adoc/bpftrace.adoc:1642]: Ah, sorry, I prefer `strcontains` since, as you said, `strstr` has a different semantics in C.

File: src/ast/irbuilderbpf.cpp
Comment (liutingjieni) [reply src/ast/irbuilderbpf.cpp]: sorry, I don't understand what you mean. here I just mimic the implementation of strncmp.

Comment (viktormalik) [reply src/ast/irbuilderbpf.cpp]: If both strings are literal, you can directly determine the result here (e.g. by using `std::string::find`) and return it.

Comment (viktormalik) [reply src/ast/irbuilderbpf.cpp]: Yes, it could use the same optimization.

**Comment by liutingjieni** [comment]
> Plus there are several unresolved comments (and some that were incorrectly resolved) from the previous reviews.

Are there any incorrectly resolved now?

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [inline src/ast/irbuilderbpf.cpp:867]: What about the other option, when `s2` is not a substring?

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [inline src/ast/irbuilderbpf.cpp:874]: Isn't this way too complicated? Can't we just do:
```
return getInt64(0);
```

File: man/adoc/bpftrace.adoc
Comment (viktormalik) [inline man/adoc/bpftrace.adoc:1642]: Looking at this now, why return 0 when the substring is found? I'd expect a function called `strcontains` returns true if it succeeds (so that you can test for the positive case, i.e. `if (strcontains(s1, s2)) ...`

**Review by viktormalik — COMMENTED** [review]
Could you also merge the two commits and cleanup the commit message?
Thanks!

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [inline src/ast/irbuilderbpf.cpp:866]:
```suggestion
    if (position != std::string::npos)
```

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [inline src/ast/irbuilderbpf.cpp:872]:
```suggestion
      return getInt64(0);
```

**Review by viktormalik — APPROVED** [review]
Since there has been a lot of work, I'd like to have one more review before merging this.

**Comment by viktormalik** [comment]
cc'ing @danobi and @ajor explicitly since this was opened before we added the CODEOWNERS file.

**Comment by xh4n3** [comment]
Since this is similar to #2387, I guess you may need to add the codegen tests and change the `CHANELOG.md`.
And maybe a little change to the empty lines, see https://github.com/iovisor/bpftrace/pull/2387#discussion_r1040584036

File: tests/runtime/call
Comment (fbs) [inline tests/runtime/call:329]: can we have a more descriptive name for this test?

File: tests/runtime/call
Comment (fbs) [inline tests/runtime/call:335]: would also be nice to have a runtime test that just tests with a few combinations of constants to make sure that codepath works

**Review by fbs — CHANGES_REQUESTED** [review]
looks good! Just a few suggestions on the testing side

File: tests/runtime/call
Comment (liutingjieni) [reply tests/runtime/call:335]: Do you have  an example of runtime test？

Comment (fbs) [reply tests/runtime/call:335]: `RUN {{BPFTRACE}} -ve 'BEGIN { if (!strcontains("abc", "a")) { printf("OK\n"); exit(); } }'`

File: tests/runtime/call
Comment (fbs) [inline tests/runtime/call:336]: we try to keep the test name descriptive of what the test does. 

```suggestion
NAME strcontains literals
```

**Pull request merged**
